R/guide-legend.r

```diff
@@ -125,624 +125,627 @@
+    byrow = byrow,
+    reverse = reverse,
+    order = order,
+
+    # Fixed parameters
+    available_aes = "any",
+    name  = "legend",
+    super = GuideLegend
   )
 }
 
+#' @rdname ggplot2-ggproto
+#' @format NULL
+#' @usage NULL
 #' @export
-guide_train.legend <- function(guide, scale, aesthetic = NULL) {
-  breaks <- scale$get_breaks()
-  if (length(breaks) == 0 || all(is.na(breaks))) {
-    return()
-  }
+GuideLegend <- ggproto(
+  "GuideLegend", Guide,
+
+  params = list(
+    title = waiver(),
+    title.position = NULL,
+    title.theme = NULL,
+    title.hjust = NULL,
+    title.vjust = NULL,
+
+    label = TRUE,
+    label.position = NULL,
+    label.theme = NULL,
+    label.hjust = NULL,
+    label.vjust = NULL,
+
+    keywidth  = NULL,
+    keyheight = NULL,
+
+    # General
+    direction = NULL,
+    override.aes = list(),
+    nrow = NULL,
+    ncol = NULL,
+    byrow = FALSE,
+    reverse = FALSE,
+    order = 0,
+
+    name  = "legend",
+    hash  = character(),
+    position = NULL,
+    direction = NULL
+  ),
+
+  available_aes = "any",
+
+  hashables = exprs(title, key$.label, direction, name),
+
+  elements = list(
+    background  = "legend.background",
+    margin      = "legend.margin",
+    spacing     = "legend.spacing",
+    spacing.x   = "legend.spacing.x",
+    spacing.y   = "legend.spacing.y",
+    key         = "legend.key",
+    key.height  = "legend.key.height",
+    key.width   = "legend.key.width",
+    text        = "legend.text",
+    text.align  = "legend.text.align",
+    theme.title = "legend.title",
+    title.align = "legend.title.align"
+  ),
+
+  extract_params = function(scale, params, hashables,
+                            title = waiver(), direction = NULL, ...) {
+    params$title <- scale$make_title(
+      params$title %|W|% scale$name %|W|% title
+    )
+    params$direction <- arg_match0(
+      params$direction %||% direction,
+      c("horizontal", "vertical"), arg_nm = "direction"
+    )
+    if (isTRUE(params$reverse %||% FALSE)) {
+      params$key <- params$key[nrow(params$key):1, , drop = FALSE]
+    }
 
-  # in the key data frame, use either the aesthetic provided as
-  # argument to this function or, as a fall back, the first in the vector
-  # of possible aesthetics handled by the scale
-  aes_column_name <- aesthetic %||% scale$aesthetics[1]
-  key <- data_frame(scale$map(breaks), .name_repair = ~ aes_column_name)
-  key$.label <- scale$get_labels(breaks)
-
-  # Drop out-of-range values for continuous scale
-  # (should use scale$oob?)
-  if (!scale$is_discrete()) {
-    limits <- scale$get_limits()
-    noob <- !is.na(breaks) & limits[1] <= breaks & breaks <= limits[2]
-    key <- key[noob, , drop = FALSE]
-  }
+    Guide$extract_params(scale, params, hashables)
+  },
 
+  merge = function(self, params, new_guide, new_params) {
+    # Combine keys
+    new_params$key$.label <- new_params$key$.value <- NULL
+    params$key <- vec_cbind(params$key, new_params$key)
```

Comment:
Because both the new and old params have matching hashes, and `key$.label` is part of hash, we can be guaranteed the labels are in the same order. If you'd use this method an a sandbox without checking the hashes, you might be in trouble.